src/utils.ts

```diff
@@ -145,29 +146,77 @@ export const getCioClient = (apiKey?: string, cioJsClientOptions?: ConstructorCl
 
 export const getActiveSectionsWithData = (
   activeSections: UserDefinedSection[],
-  sectionResults: SectionsData
+  sectionResults: SectionsData,
+  sectionsRefs: React.MutableRefObject<React.RefObject<HTMLLIElement>[]>
 ) => {
   const activeSectionsWithData: Section[] = [];
-  activeSections?.forEach((config) => {
-    const { identifier } = config;
-    let data;
+  activeSections?.forEach((sectionConfig, index) => {
+    const { identifier } = sectionConfig;
+    let sectionData;
 
-    if (isCustomSection(config)) {
+    if (isCustomSection(sectionConfig)) {
       // Copy id from data to the top level
-      data = config.data.map((item) => ({
+      sectionData = sectionConfig.data.map((item) => ({
         ...item,
         id: item?.id || item?.data?.id,
       }));
     } else {
-      data = sectionResults[identifier];
+      sectionData = sectionResults[identifier];
     }
 
-    if (Array.isArray(data)) {
-      activeSectionsWithData.push({ ...config, data });
+    if (Array.isArray(sectionData)) {
+      const section = {
+        ...sectionConfig,
+        data: sectionData,
+      };
+
+      if (sectionConfig.type === 'recommendations') {
+        // If user provided a ref in config, use it. Otherwise, use the ref from our refs array
+        section.ref = sectionConfig.ref || sectionsRefs.current[index];
+      }
+
+      activeSectionsWithData.push(section);
     }
   });
 
   return activeSectionsWithData;
 };
 
 export const escapeRegExp = (string: string) => string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
+
+export const trackRecommendationView = (
+  target: HTMLElement,
+  activeSectionsWithData: Section[],
+  cioClient: Nullable<ConstructorIOClient>
+) => {
+  if (target.dataset.cnstrcRecommendationsPodId) {
+    // Pull recommendations from activeSectionsWithData by podId surfaced on target
+    const recommendationSection = activeSectionsWithData.find(
+      (section) => section.identifier === target.dataset.cnstrcRecommendationsPodId
+    );
+    const recommendationItems = recommendationSection?.data.map((item) => ({
+      itemId: item.data?.id,
```

Comment:
Based on the defined types It should be an object and at this step it should always be defined. So if it's not an object or if it's undefined this means that there's a bug that needs to be addressed. Based on the previous I believe it's better not to add optional chaining since we would need this to throw an error if it's undefined or not an object so we can catch it